### PR TITLE
Implement functions with zero return values

### PR DESCRIPTION
**Pull request by hugomg**
I think functions that return nothing are better "void" functions than functions that return "nil".

This also changes the test for "control reaches end of non-void" function to require a "return nil" for functions that return nil, and only make the return statement be optional in functions that return 0 values.

### CODE REVIEW

File: spec/checker_spec.lua [hunk 1]
```diff
@@ -688,10 +728,24 @@ describe("Titan type checker", function()
...
+            local function f(): ()
+            end
+
+            local function g(): integer
+                local x = 1 + f()
+                return x
+            end
+        ]])
+        assert.falsy(prog)
+        assert.match("expected integer but found invalid type", errs)
```
Comment (gligneul) [inline spec/checker_spec.lua]: Maybe we should add an `void` type? I think this message would be clearer.

File: titan-compiler/checker.lua [hunk 2]
```diff
@@ -112,7 +112,7 @@ check_type = function(typ, errors)
         return types.T.Array(check_type(typ.subtype, errors))
 
     elseif tag == ast.Type.Function then
-        if #typ.argtypes ~= 1 then
+        if #typ.rettypes >= 2 then
             error("functions with 0 or 2+ return values are not yet implemented")
```
Comment (gligneul) [inline titan-compiler/checker.lua]: We should update the assert message to `"function with 2+ return...`

File: titan-compiler/checker.lua [hunk 3]
```diff
@@ -147,7 +147,7 @@ check_toplevel = function(tl_node, errors)
         end
 
     elseif tag == ast.Toplevel.Func then
-        if #tl_node.rettypes ~= 1 then
+        if #tl_node.rettypes >= 2 then
             error("functions with 0 or 2+ return values are not yet implemented")
```
Comment (gligneul) [inline titan-compiler/checker.lua]: Same here.

File: titan-compiler/checker.lua [hunk 4]
```diff
@@ -644,8 +653,12 @@ check_exp = function(exp, errors, typehint)
                     "function %s called with %d arguments but expects %d",
                     fname, nargs, nparams)
             end
-            assert(#ftype.rettypes == 1)
-            exp._type = ftype.rettypes[1]
+            assert(#ftype.rettypes <= 1)
+            if #ftype.rettypes >= 1 then
+                exp._type = ftype.rettypes[1]
+            else
+                exp._type = types.T.Invalid()
```
Comment (gligneul) [inline titan-compiler/checker.lua]: types.T.Void()?

File: titan-compiler/coder.lua [hunk 5]
```diff
@@ -337,23 +342,41 @@ generate_program = function(prog, modname)
...
+                    set_ret = util.render([[
+                        ${SET_SLOT}
+                        api_incr_top(L);
+                    ]], {
+                        SET_SLOT = set_slot(ret_typ, "s2v(L->top)", "ret")
+                    })
+                else
+                    error("not implemented")
+                end
+
```
Comment (gligneul) [inline titan-compiler/coder.lua:361]: s/\n//

File: titan-compiler/coder.lua [hunk 6]
```diff
@@ -337,23 +342,41 @@ generate_program = function(prog, modname)
                     table.insert(args, arg_name)
                 end
 
+                local ret_init, set_ret
+                if #tl_node._type.rettypes == 0 then
+                    ret_init = ""
+                    set_ret = ""
+                elseif #tl_node._type.rettypes == 1 then
+                    local ret_typ = tl_node._type.rettypes[1]
+                        ret_init = c_declaration(ctype(ret_typ), "ret") .. " ="
```
Comment (gligneul) [inline titan-compiler/coder.lua]: There is an extra indentation level here.

File: titan-compiler/coder.lua [hunk 7]
```diff
@@ -627,16 +650,26 @@ generate_stat = function(stat)
...
+            ${IGNORE_RESULT}
         ]], {
             STATS = cstats,
-            VALUE = cvalue
+            IGNORE_RESULT = ignore_result
         })
 
     elseif tag == ast.Stat.Return then
-        local cstats, cvalue = generate_exp(stat.exp)
+        assert(#stat.exps == 1)
```
Comment (gligneul) [inline titan-compiler/coder.lua]: What happens if we have a return with no exps (eg. void function)?

File: titan-compiler/coder.lua [hunk 8]
```diff
@@ -804,20 +837,27 @@ generate_exp = function(exp) -- TODO
...
                 local cstats = util.render([[
                     ${ARG_STATS}
-                    ${TMP_DECL} = ${FUN_NAME}(${ARGS});
+                    ${TMP_INIT} ${FUN_NAME}(${ARGS});
                 ]], {
                     FUN_NAME  = tl_node._titan_entry_point,
                     ARG_STATS = table.concat(arg_cstatss, "\n"),
                     ARGS      = table.concat(arg_cvalues, ", "),
-                    TMP_DECL = tmp_decl,
+                    TMP_INIT = tmp_init,
```
Comment (gligneul) [inline titan-compiler/coder.lua]: missing a space before `=`

**Review by gligneul — CHANGES_REQUESTED** [review]
Just minor stuff and some questions

**Review by gligneul — APPROVED** [review]
